From: YAOF native build <builder@localhost>
Subject: [PATCH] metadata: keep provider rank out of re-emitted dependencies

Backport the rank/negation separation from OpenWrt main commit
3c3a4150694ca9ea4b8a44c12273bd4780bbf7d0. Preserve tristate provider
selection while avoiding consumer self-references in propagated dependencies.

--- a/scripts/package-metadata.pl
+++ b/scripts/package-metadata.pl
@@ -180,6 +180,8 @@
 		}
 		if ($flags =~ /\+/) {
 			my $vdep = $vpackage{$depend};
+			my $rank;
+			my $negation;
 			if ($vdep) {
 				my @vdeps;
 
@@ -195,22 +197,31 @@
 				$depend = shift @vdeps;
 
 				if (@vdeps > 1) {
-					$condition = ($condition ? "$condition && " : '') . join("&&", map { "PACKAGE_$_<PACKAGE_$pkgname" } @vdeps);
+					$rank = join("&&", map { "PACKAGE_$_<PACKAGE_$pkgname" } @vdeps);
+					$negation = '!('.join("||", map { "PACKAGE_$_" } @vdeps).')';
 				} elsif (@vdeps > 0) {
-					$condition = ($condition ? "$condition && " : '') . "PACKAGE_${vdeps[0]}<PACKAGE_$pkgname";
+					$rank = "PACKAGE_${vdeps[0]}<PACKAGE_$pkgname";
+					$negation = '!PACKAGE_'.$vdeps[0];
 				}
 			}
 
 			# Menuconfig will not treat 'select FOO' as a real dependency
 			# thus if FOO depends on other config options, these dependencies
 			# will not be checked. To fix this, we simply emit all of FOO's
-			# depends here as well.
-			$package{$depend} and push @t_depends, [ $package{$depend}->{depends}, $condition ];
+			# depends here without the consumer's provider rank.
+			my $reemit = $condition;
+			$reemit = ($reemit ? "$reemit && " : '') . $negation if defined $negation;
+			$package{$depend} and push @t_depends, [ $package{$depend}->{depends}, $reemit ];
 
 			$m = "select";
 			next if $only_dep;
 
 			$flags =~ /@/ or $depend = "PACKAGE_$depend";
+
+			# Provider rank applies only to select emission.
+			if (defined $rank) {
+				$condition = ($condition ? "$condition && " : '') . $rank;
+			}
 		} else {
 			my $vdep = $vpackage{$depend};
 			if ($vdep && @$vdep > 0) {
